Dashboard Enhancement - Support Drag, Drop and Resize

ns8482e opened the pull request:
Added support for drag drop sort and resize
Fixes #8601

![dashboard-3](https://user-images.githubusercontent.com/23270244/108021860-0a8ae200-6fe5-11eb-98d0-05124601df3f.gif)

hishamco commented:
Great work Niraj, can we have a demo in the standup?

agriffard commented:
![image](https://user-images.githubusercontent.com/703248/108043009-d9d89600-7040-11eb-8813-10bb6a05abf0.png)

Here is how it looks like when I enable the feature.
The bottom is over the border probably because of the `list-group-flush` class in the list created from the recipe.

Could we hide the date and status in View mode?

When I click on 'Manage dashboard':

![image](https://user-images.githubusercontent.com/703248/108043247-2de37a80-7041-11eb-96bc-19dbb30b8746.png)

Add some space between buttons.
I didn't understand what was doing the 'View' button until I click on it. It is more like a 'Go back to dashboard'.

When I create or edit a widget and specify Height = 2 or Width = 2, it is not taken into account.

![image](https://user-images.githubusercontent.com/703248/108045422-bbc06500-7043-11eb-8775-2364e0aba7ea.png)

Add space between` successfully updated.` and `Undo`

agriffard commented:
If possible: Do not display Card header if title is empty.

![image](https://user-images.githubusercontent.com/703248/108050229-a2221c00-7049-11eb-9eec-6fe99bd75fb7.png)

ns8482e commented:
> If possible: Do not display Card header if title is empty.

Good catch - seems an issue with `TitlePart`, will fix it.

deanmarcussen commented on src/OrchardCore.Modules/OrchardCore.AdminDashboard/DashboardManageAttribute.cs:
```diff
@@ -0,0 +1,22 @@
+namespace OrchardCore.AdminDashboard
+{
+    [AttributeUsage(AttributeTargets.Class | AttributeTargets.Method, AllowMultiple = false, Inherited = true)]
+    public class DashboardManageAttribute : Attribute, IAsyncResourceFilter
```
I would suggest just using an `httpContext.Feature` rather than needing another filter here.

ns8482e replied:
Updated code to use Feature

deanmarcussen commented on src/OrchardCore.Modules/OrchardCore.AdminDashboard/Drivers/DashboardContentDisplayDriver.cs:
```diff
@@ -0,0 +1,119 @@
+        {
+            var httpContext = _httpContextAccessor.HttpContext;
+            var isManageRequest = httpContext.Items.ContainsKey(typeof(DashboardManageAttribute));
+            var hasPublished = await _contentManager.HasPublishedVersionAsync(model);
```
Here I think you can just return `Task.FromResult<IDisplayResult>(null)` if the feature is not set.

ns8482e replied:
Done

deanmarcussen commented on src/OrchardCore.Modules/OrchardCore.AdminDashboard/Drivers/DashboardContentDisplayDriver.cs:
```diff
@@ -0,0 +1,119 @@
+            {
+                m.ContentItem = model;
+            }).Location("ActionsMenu:after")
+            .RenderWhen(async () =>
```
Probably you don't need to use `RenderWhen` here at all.

If you just return null on `!isManageRequest`

ns8482e replied:
Removed `RenderWhen`

deanmarcussen commented on src/OrchardCore.Modules/OrchardCore.AdminDashboard/Drivers/DashboardContentDisplayDriver.cs:
```diff
@@ -0,0 +1,119 @@
+            {
+                
+                return await Task.FromResult(isManageRequest && hasDeletePermission);
+            });
```
Here you can just return the Task, rather than awaiting a result.
i.e.
`.RenderWhen(() => Task.FromResult(hasDeletePermission)`

ns8482e replied:
Removed `RenderWhen`

deanmarcussen commented on src/OrchardCore.Modules/OrchardCore.AdminDashboard/Drivers/DashboardContentDisplayDriver.cs:
```diff
@@ -0,0 +1,119 @@
+            });
+            results.Add(shapeTag);
+
+            return await Task.FromResult<IDisplayResult>(Combine(results.ToArray()));
```
```suggestion
            return await Combine(results.ToArray());
```

deanmarcussen commented on src/OrchardCore.Modules/OrchardCore.Html/Drivers/HtmlBodyPartDisplayDriver.cs:
```diff
@@ -43,7 +43,6 @@ public override IDisplayResult Display(HtmlBodyPart HtmlBodyPart, BuildPartDispl
             return Initialize<HtmlBodyPartViewModel>(GetDisplayShapeType(context), m => BuildViewModelAsync(m, HtmlBodyPart, context))
                 .Location("Detail", "Content:5")
                 .Location("Summary", "Content:10")
-                .Location("DetailAdmin", "Content:10") // For dashboard widgets;
```
Intentional to remove these?

Why?

ns8482e replied:
Reusing existing `Detail` `DisplayType` and it's placement, so no need of  defining `DetailAdmin` Display type and it's placement.

sebastienros replied:
It can't be Detail. Detail is for front-end.

ns8482e replied:
UI content is same for `Detail` and `DetailAdmin`, why create new `DisplayType`?  Creating `DetailAdmin` display type means we need to define `Location` in drivers for every parts available in `OrchardCore` and duplicate the templates with same code

ns8482e replied:
Revert back to use `DetailAdmin`

ns8482e replied:
@dalenewman  @sebastienros  with `DetailAdmin` - display type

Currently adding a field or part (e.g. media field) in dashboard widget doesn't render them by default and requires to manually define placement for each field or part used

sebastienros replied:
I think we realized that with @agriffard and the solution is to add a default placement for DetailAdmin like we do for Detail.

sebastienros replied:
Apparently like the one you are removing here ;)

deanmarcussen commented on src/OrchardCore.Modules/OrchardCore.Markdown/Drivers/MarkdownBodyPartDisplayDriver.cs:
```diff
@@ -47,7 +47,6 @@ public override IDisplayResult Display(MarkdownBodyPart markdownBodyPart, BuildP
             return Initialize<MarkdownBodyPartViewModel>(GetDisplayShapeType(context), m => BuildViewModel(m, markdownBodyPart, context))
                 .Location("Detail", "Content:10")
                 .Location("Summary", "Content:10")
-                .Location("DetailAdmin", "Content:10") // For dashboard widgets
```
as above

ns8482e replied:
Reusing existing Detail DisplayType and it's placement, so no need of defining DetailAdmin Display type and it's placement.

deanmarcussen replied:
I think the intention here was actually to have a different display type, so it could be customised for the admin if required.

ns8482e replied:
Revert back to use `DetailAdmin`

deanmarcussen reviewed (commented):
For info @ns8482e 

The `.RenderWhen` is a small optimization to prevent too much work being done when items aren't being placed. (Like when rendering `Detail` so stopping work that `SummaryAdmin` shapes have to do, (mostly for auth) that will never be placed).

So not so necesary here, as you can just get out quicker if it's not dashboard widget related.

sebastienros commented:
I don't see the reason to display the metadata when not in edit mode. Most of the time this dashboard will be in display mode, so this should be optimized for this mode.

agriffard commented:
A `dropdown-item` class is missing on some buttons like Unpublish.

ns8482e commented:
> I don't see the reason to display the metadata when not in edit mode. Most of the time this dashboard will be in display mode, so this should be optimized for this mode.

Removed metadata from View

ns8482e commented:
> A `dropdown-item` class is missing on some buttons like Unpublish.

fixed

agriffard commented:
Can you please enahnce the drag and drop area?
The yellow triangle is difficult to understand.
Can you display a zone with a dashed border?

ns8482e commented:
> Can you please enahnce the drag and drop area?
> The yellow triangle is difficult to understand.
> Can you display a zone with a dashed border?

- That could be because of dense flow, I removed it for now, may be we could add it later in settings.
- Removed triangle drop cursor and replaced with border.

deanmarcussen commented on src/OrchardCore.Modules/OrchardCore.AdminDashboard/Drivers/DashboardContentDisplayDriver.cs:
```diff
@@ -0,0 +1,109 @@
+            var hasDeletePermission = await _authorizationService.AuthorizeAsync(httpContext.User, CommonPermissions.DeleteContent, model);
+            var hasPublishPermission = await _authorizationService.AuthorizeAsync(httpContext.User, CommonPermissions.PublishContent, model);
+
+            var dragHandle = Dynamic("Dashboard_DragHandle", m =>
```
```suggestion
            var dragHandle = Initialize<ContentItemViewModel>("Dashboard_DragHandle", m =>
```

Great thanks for moving it to a feature, and cleaning up the async.

Suggestion here is that we are trying to move away from dynamic where possible, and of all the shape factory creations, this is the better one for allocations. It will create a strongly typed `ContentItemViewModel` for each, which means you can use that as the model in the views, rather than dynamic.

I'm not seeing any reason for them to be dynamic, unless I am wrong?

So here, and all the other shapes from this driver, look like they can be initialized.

Can you also have a look through and clean up all the extra line breaks too please ?

ns8482e replied:
done

deanmarcussen commented on src/OrchardCore.Modules/OrchardCore.AdminDashboard/Drivers/DashboardContentDisplayDriver.cs:
```diff
@@ -0,0 +1,109 @@
+                }).Location("ActionsMenu:after");
+                results.Add(discardDraftButton);
+            }
+            var shapeTag = Shape("DashboardWidget_DetailAdmin__ContentsTags", new ContentItemViewModel(model))
```
Here to `Initialize<ContentItemViewModel>` is better

ns8482e replied:
done

sebastienros commented:
Some other comments from the meeting that happened yesterday and is not yet live:
- Would be great to see the "snap" effect on the grid when we resize
- Hide the position/width/height inputs behind a type setting such that by default they are not displayed when we edit a widget. The visual editor is sufficient now.
- Rename the "View" button into "Close", and "New" into "Add". Could be "Close Editor", "Leave Editor" something more obvious. And also "Add Widget" if there is space. I found "New" and "View" not obvious in their intent.

sebastienros commented:
We need to start thinking about which widgets we should provide in some recipes (not the scope of this PR). For instance the Blog recipe could enable it, and provide a widget displaying the latest posts, archive stats, tags stats (tag cloud?). This could be done with dynamic types and liquid templates, and would be a great sandbox for devs to see what can be done with queries and liquid. A good set of ideas will probably be found in WordPress.
Alternatively each module could provide at least one widget type.

ns8482e commented:
> Some other comments from the meeting that happened yesterday and is not yet live:
> 
> * Would be great to see the "snap" effect on the grid when we resize

Done

> * Hide the position/width/height inputs behind a type setting such that by default they are not displayed when we edit a widget. The visual editor is sufficient now.

Does it mean we don't need `DashboardPartDisplayDriver` and attach `DashboardPart` using weld - the magical way and not needed to add `DashboardPart` manually?

> * Rename the "View" button into "Close", and "New" into "Add". Could be "Close Editor", "Leave Editor" something more obvious. And also "Add Widget" if there is space. I found "New" and "View" not obvious in their intent.

Done

> * The grid should always be 6 of width, and not switch to 7 when the menu is collapsed

Currently it's responsive - adding fix width, will also apply to smaller screen and become non-responsive. We can add configurable columns/row settings in future PR.

giannik commented:
#8769
